Library/Homebrew/cmd/update.sh

```diff
@@ -378,19 +381,34 @@ EOS
   for DIR in "$HOMEBREW_REPOSITORY" "$HOMEBREW_LIBRARY"/Taps/*/*
   do
     [[ -d "$DIR/.git" ]] || continue
+    cd "$DIR" || continue
+
+    TAP_VAR="$(repo_var "$DIR")"
+    declare PREFETCH_REVISION"$TAP_VAR"="$(read_current_revision)"
```

Comment:
After a night's sleep and reviewing the changes, I think we might have to make some more tweaks to the logic changed here. 😢 The fix isn't actually fixing the problem, though it's improving the situation to a point where people won't be indefinitely stuck:

```console
$ cd "$(brew --repository)"
$ git checkout master
$ git fetch origin
$ git reset --hard 985c672b # move back to fix commit (not up-to-date with upstream)
HEAD is now at 985c672b update.sh: check upstream SHA prefetch not local.
$ git push --force . 985c672b:origin/master # pretend we didn't fetch the latest commit
Total 0 (delta 0), reused 0 (delta 0)
To .
 + b9d5b11f...985c672b 985c672b -> origin/master (forced update)
$ brew update
Already up-to-date.
$ git rev-parse master
985c672bac4dc20d369b451c484eb6553762dbcf
$ git rev-parse origin/master
b9d5b11f321a359b462aa2c353c2bdb483930fb3
```

Note how `brew update` is claiming up-to-date, but `master` hasn't been updated to `origin/master`. A second `brew update` will actually update, but with the current fix every user will have to run `brew update` twice in a row to actually get updated.

It might also be doing unintended things if something different from `master` is checked out, as it's comparing whatever is currently checked out to `origin/master` (but before any new commits have been fetched and the `origin/master` ref updated).